ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2835,194 @@ public sealed class Bot : IAsyncDisposable {
+			return appIDs;
+		}
+
+		private async Task<HashSet<uint>> GetPossiblyCompletedBadgeAppIDs(byte page) {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(page).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return new HashSet<uint>(0);
+			}
+
+			return GetPossiblyCompletedBadgeAppIDs(badgePage);
+		}
+
+		private HashSet<uint> GetPossiblyCompletedBadgeAppIDs(IDocument badgePage) {
+			List<IElement> linkElements = badgePage.SelectNodes("//a[@class='badge_craft_button']");
+
+			// We need to also select all badges that we have max level, as those will not display with a craft button
+			// Level 5 is maximum level for card badges according to https://steamcommunity.com/tradingcards/faq
+			linkElements.AddRange(badgePage.SelectNodes("//div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					continue;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Replace("?border=1", "").Split('/', StringSplitOptions.RemoveEmptyEntries).Last();
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					continue;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private async Task SendCompletedSets() {
+			ISet<uint> appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+			if (appIDs.Count == 0) {
+				// Could not extract appIds for craftable badges - No need to load the inventory at this point
+				return;
+			}
+
+			HashSet<Steam.Asset> inventory;
+
+			try {
+				inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+					.Where(item => item.Tradable)
+					.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+					.Where(item => appIDs.Contains(item.RealAppID))
+					.ToHashSetAsync()
+					.ConfigureAwait(false);
+			} catch (HttpRequestException e) {
+				ArchiLogger.LogGenericWarningException(e);
+
+				return;
+			} catch (Exception e) {
+				ArchiLogger.LogGenericException(e);
+
+				return;
+			}
+
+			if (inventory.Count == 0) {
+				ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+				return;
+			}
+
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), uint> inventorySets = Trading.GetInventorySets(inventory).ToDictionary(entry => entry.Key, entry => entry.Value.FirstOrDefault());
```

Comment:
You're throwing here away so many CPU cycles and memory, creating new dictionary from already existing dictionary only because you're too damn lazy to take the first entry from the `List<uint>`? REALLY?